File: python/astal/variable.py
from gi.repository import Astal

from .binding import Binding


class Variable:
    def __init__(self, init):
        v = Astal.Variable.new(init)
        self._variable = v
        self._err_handler = print
        v.connect("error", lambda _, err: self._err_handler(err) if self._err_handler else None)

    def __call__(self, transform=None):
        if transform:
            return Binding(self).as_(transform)

        return Binding(self)

    def __str__(self):
        return f"Variable<{self.get()}>"

    def get(self):
        return self._variable.get_value()

    def set(self, value):
        return self._variable.set_value(value)

    def watch(self, cmd):
        if isinstance(cmd, str):
            self._variable.watch(cmd)
        elif isinstance(cmd, list):
            self._variable.watchv(cmd)
        return self

    def poll(self, interval, cmd):
        if isinstance(cmd, str):
            self._variable.poll(interval, cmd)
        elif isinstance(cmd, list):
            self._variable.pollv(interval, cmd)
        else:
            self._variable.pollfn(interval, cmd)
        return self

    def start_watch(self):
        self._variable.start_watch()

    def start_poll(self):
        self._variable.start_poll()

    def stop_watch(self):
        self._variable.stop_watch()

    def stop_poll(self):
        self._variable.stop_poll()

    def drop(self):
        self._variable.emit_dropped()
        self._variable.run_dispose()

    def on_dropped(self, callback):
        self._variable.connect("dropped", lambda _: callback())
        return self

    def on_error(self, callback):
        self._err_handler = None
        self._variable.connect("error", lambda _, e: callback(e))
        return self

    def subscribe(self, callback):
        s = self._variable.connect("changed", lambda _: callback(self.get()))
        return lambda: self._variable.disconnect(s)

    def observe(self, objs, sigOrFn, callback=None):
        if callable(sigOrFn):
            f = sigOrFn
        elif callable(callback):
            f = callback
        else:
            f = lambda *_: self.get()

        def setter(*args):
            self.set(f(*args))

        if isinstance(objs, list):
            for obj in objs:
                obj[0].connect(obj[1], setter)
        elif isinstance(sigOrFn, str):
            objs.connect(sigOrFn, setter)

        return self

    @staticmethod
    def derive(deps, fn):
        def update():
            return fn(*[d.get() for d in deps])

        derived = Variable(update())
        unsubs = [dep.subscribe(lambda _: derived.set(update())) for dep in deps]
        derived.on_dropped(lambda: ([unsub() for unsub in unsubs]))
        return derived

File: lang/gjs/src/variable.ts
import Astal from "gi://AstalIO"
import Binding, { type Connectable, type Subscribable } from "./binding.js"
import { interval } from "./time.js"
import { execAsync, subprocess } from "./process.js"

class VariableWrapper<T> extends Function {
    private variable!: Astal.VariableBase
    private errHandler? = console.error

    private _value: T
    private _poll?: Astal.Time
    private _watch?: Astal.Process

    private pollInterval = 1000
    private pollExec?: string[] | string
    private pollTransform?: (stdout: string, prev: T) => T
    private pollFn?: (prev: T) => T | Promise<T>

    private watchTransform?: (stdout: string, prev: T) => T
    private watchExec?: string[] | string

    constructor(init: T) {
        super()
        this._value = init
        this.variable = new Astal.VariableBase()
        this.variable.connect("dropped", () => {
            this.stopWatch()
            this.stopPoll()
        })
        this.variable.connect("error", (_, err) => this.errHandler?.(err))
        return new Proxy(this, {
            apply: (target, _, args) => target._call(args[0]),
        })
    }

    private _call<R = T>(transform?: (value: T) => R): Binding<R> {
        const b = Binding.bind(this)
        return transform ? b.as(transform) : b as unknown as Binding<R>
    }

    toString() {
        return String(`Variable<${this.get()}>`)
    }

    get(): T { return this._value }
    set(value: T) {
        if (value !== this._value) {
            this._value = value
            this.variable.emit("changed")
        }
    }

    startPoll() {
        if (this._poll)
            return

        if (this.pollFn) {
            this._poll = interval(this.pollInterval, () => {
                const v = this.pollFn!(this.get())
                if (v instanceof Promise) {
                    v.then(v => this.set(v))
                        .catch(err => this.variable.emit("error", err))
                }
                else {
                    this.set(v)
                }
            })
        }
        else if (this.pollExec) {
            this._poll = interval(this.pollInterval, () => {
                execAsync(this.pollExec!)
                    .then(v => this.set(this.pollTransform!(v, this.get())))
                    .catch(err => this.variable.emit("error", err))
            })
        }
    }

    startWatch() {
        if (this._watch)
            return

        this._watch = subprocess({
            cmd: this.watchExec!,
            out: out => this.set(this.watchTransform!(out, this.get())),
            err: err => this.variable.emit("error", err),
        })
    }

    stopPoll() {
        this._poll?.cancel()
        delete this._poll
    }

    stopWatch() {
        this._watch?.kill()
        delete this._watch
    }

    isPolling() { return !!this._poll }
    isWatching() { return !!this._watch }

    drop() {
        this.variable.emit("dropped")
    }

    onDropped(callback: () => void) {
        this.variable.connect("dropped", callback)
        return this as unknown as Variable<T>
    }

    onError(callback: (err: string) => void) {
        delete this.errHandler
        this.variable.connect("error", (_, err) => callback(err))
        return this as unknown as Variable<T>
    }

    subscribe(callback: (value: T) => void) {
        const id = this.variable.connect("changed", () => {
            callback(this.get())
        })
        return () => this.variable.disconnect(id)
    }

    poll(
        interval: number,
        exec: string | string[],
        transform?: (stdout: string, prev: T) => T
    ): Variable<T>

    poll(
        interval: number,
        callback: (prev: T) => T | Promise<T>
    ): Variable<T>

    poll(
        interval: number,
        exec: string | string[] | ((prev: T) => T | Promise<T>),
        transform: (stdout: string, prev: T) => T = out => out as T,
    ) {
        this.stopPoll()
        this.pollInterval = interval
        this.pollTransform = transform
        if (typeof exec === "function") {
            this.pollFn = exec
            delete this.pollExec
        }
        else {
            this.pollExec = exec
            delete this.pollFn
        }
        this.startPoll()
        return this as unknown as Variable<T>
    }

    watch(
        exec: string | string[],
        transform: (stdout: string, prev: T) => T = out => out as T,
    ) {
        this.stopWatch()
        this.watchExec = exec
        this.watchTransform = transform
        this.startWatch()
        return this as unknown as Variable<T>
    }

    observe(
        objs: Array<[obj: Connectable, signal: string]>,
        callback: (...args: any[]) => T,
    ): Variable<T>

    observe(
        obj: Connectable,
        signal: string,
        callback: (...args: any[]) => T,
    ): Variable<T>

    observe(
        objs: Connectable | Array<[obj: Connectable, signal: string]>,
        sigOrFn: string | ((obj: Connectable, ...args: any[]) => T),
        callback?: (obj: Connectable, ...args: any[]) => T,
    ) {
        const f = typeof sigOrFn === "function" ? sigOrFn : callback ?? (() => this.get())
        const set = (obj: Connectable, ...args: any[]) => this.set(f(obj, ...args))

        if (Array.isArray(objs)) {
            for (const obj of objs) {
                const [o, s] = obj
                const id = o.connect(s, set)
                this.onDropped(() => o.disconnect(id))
            }
        }
        else {
            if (typeof sigOrFn === "string") {
                const id = objs.connect(sigOrFn, set)
                this.onDropped(() => objs.disconnect(id))
            }
        }

        return this as unknown as Variable<T>
    }

    static derive<
        const Deps extends Array<Subscribable<any>>,
        Args extends {
            [K in keyof Deps]: Deps[K] extends Subscribable<infer T> ? T : never
        },
        V = Args,
    >(deps: Deps, fn: (...args: Args) => V = (...args) => args as unknown as V) {
        const update = () => fn(...deps.map(d => d.get()) as Args)
        const derived = new Variable(update())
        const unsubs = deps.map(dep => dep.subscribe(() => derived.set(update())))
        derived.onDropped(() => unsubs.map(unsub => unsub()))
        return derived
    }
}

export interface Variable<T> extends Omit<VariableWrapper<T>, "bind"> {
    <R>(transform: (value: T) => R): Binding<R>
    (): Binding<T>
}

export const Variable = new Proxy(VariableWrapper as any, {
    apply: (_t, _a, args) => new VariableWrapper(args[0]),
}) as {
    derive: typeof VariableWrapper["derive"]
    <T>(init: T): Variable<T>
    new<T>(init: T): Variable<T>
}

export default Variable
